### PR TITLE
Reduce impact of stringex gem

### LINKED ISSUE #8724: Remove/replace stringex gem
### What we should change and why (this is tech debt)

We're using an old gem called `stringex` which I think is a hangover from Spree. As far as I can tell it's used in a fairly minor way in the creation of permalinks for things like enterprises. **It's got to go** :fire::fire::fire:

**The tricky part:** I can't really tell the full extent to which it's used, and where, and why...

**The bad part:** the gem works by wrecklessly monkey-patching core Ruby classes and hacking extra methods into them, including the `String` class. 

Apologies to any devs who were drinking a glass of water whilst reading that last sentence, presumably you'll need a new keyboard now.

### Context

Discovered this whilst debugging something else, and pulling on some threads.

### Impact and timeline

This gem is **bad news bears**, it's a leaky abstraction that can easily clash with other gems, with unexpected consequences. We currently need a small workaround in the split checkout code to avoid a crazy bug that's essentially caused by this gem.

It should be fairly simple to rip it out and throw it away. We could even cheat and grep some Spree branches to see what _they_ replaced it with.

### PR DESCRIPTION
**Pull request by Matt-Yorkley**
#### What? Why?

Closes #8724 

- Ensures the `stringex` gem is not loaded by default and the `String` class is no longer monkey-patched
- Replaces use of StringEx's `#to_url` with a new encapsulated method in the `UrlGenerator` service
- Adds ability to generate valid Enterprise permalinks from non-roman characters 

**Review Notes:** the built-in `ActiveSupport#parameterize` method provides 99% of the functionality provided by `StringEx#to_url`. The only thing it doesn't do is handle transliteration of non-roman characters. _That_ part of the `stringex` gem can be accessed independently via the `StringEx::UriDecode` class without loading the entire library (and without the crazy monkey-patching on `String`).

FYI we have some (pre-existing) test coverage on this transliteration functionality already:
https://github.com/openfoodfoundation/openfoodnetwork/blob/6f90a7b07fb9fd9b9afe067b751f229499e778ea/spec/models/spree/taxon_spec.rb#L64-L68

#### What should we test?

Creation of permalinks for things like products and enterprises. 

#### Release notes

Adjusted use of stringex gem for permalink generation

Changelog Category: Technical changes

### CODE REVIEW

**Review by jibees — APPROVED** [review]
Looks Good To Me!
Thanks!

File: spec/services/url_generator_spec.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+describe UrlGenerator do
+  subject { UrlGenerator }
+
+  describe "#to_url" do
+    it "converts to url-safe strings and removes unusable characters" do
+      expect(subject.to_url("Top Cat!?")).to eq "top-cat"
+    end
+
+    it "handles unusual characters like accents and umlauts" do
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply spec/services/url_generator_spec.rb:13]:
```suggestion
    it "handles accents" do
```

**Review by mkllnk — COMMENTED** [review]
Excellent! :fire: 

But I agree that the anglo-centric spec description should be changed.

**Comment by filipefurtad0** [comment]
Hey @Matt-Yorkley ,

Tested by:

- verifying availability of enterprise permalinks - OK

Peek 2022-01-26 18-41

- Created products with the same name - OK
expected behavior is to add `-1` to the existing permalink

I think we're good.
Merging.

**Pull request merged**